Forge Networking Remastered Unity/Assets/Bearded Man Studios Inc/Scripts/Networking/Forge/Networking/FacepunchP2PServer.cs

```diff
@@ -0,0 +1,752 @@
+﻿#if FACEPUNCH_STEAMWORKS
+using BeardedManStudios.Forge.Networking.Frame;
+using BeardedManStudios.Threading;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Steamworks;
+
+namespace BeardedManStudios.Forge.Networking
+{
+	public class FacepunchP2PServer : BaseFacepunchP2P, IServer
+	{
+		public List<string> BannedAddresses { get; set; }
+
+		/// <summary>
+		/// Used to determine if this server is currently accepting connections
+		/// </summary>
+		public bool AcceptingConnections { get; private set; }
+
+		public Dictionary<SteamId, SteamNetworkingPlayer> steamPlayers = new Dictionary<SteamId, SteamNetworkingPlayer>();
+
+		/// <summary>
+		/// Callback on server Host() completion
+		/// </summary>
+		public event BaseNetworkEvent serverCreated;
+
+		/// <summary>
+		/// Callback on server Host() completion
+		/// </summary>
+		protected void OnServerCreated()
+		{
+			if (serverCreated != null)
+			{
+				serverCreated(this);
+			}
+		}
+
+		protected List<FrameStream> bufferedMessages = new List<FrameStream>();
+
+		private CommonServerLogic commonServerLogic;
+		private SteamNetworkingPlayer currentReadingPlayer = null;
+
+#region Steam P2P Callbacks
+
+		/// <summary>
+		/// Callback for SteamNetworking.OnP2PSessionRequest
+		/// Accepts all incoming connections
+		/// </summary>
+		/// <param name="requestorSteamId">Incoming P2P request</param>
+		private void OnP2PSessionRequest(SteamId requestorSteamId)
+		{
+			if (AcceptingConnections)
+			{
+				if (!SteamNetworking.AcceptP2PSessionWithUser(requestorSteamId))
+					Logging.BMSLog.LogWarning("Could not accept P2P Session with user: " + requestorSteamId.Value);
+			}
+			else
+				Logging.BMSLog.LogWarning("P2P Request received but server is not accepting connections");
+		}
+
+		/// <summary>
+		/// Callback for SteamNetworking.OnP2PConnectionFailed
+		/// </summary>
+		/// <param name="remoteSteamId">SteamId of the remote peer</param>
+		private void OnP2PConnectionFailed(SteamId remoteSteamId)
+		{
+			Logging.BMSLog.Log("OnP2PConnectionFailed called. Remote steamId: " + remoteSteamId.Value.ToString());
```

Comment:
Same as on the client, maybe it would be good to have an event be fired here instead of the log?